include/swift/ABI/TaskStatus.h

```diff
@@ -154,12 +162,97 @@ class ChildTaskStatusRecord : public TaskStatusRecord {
   }
 };
 
+/// A status record which states that a task has a task group.
+///
+/// The child tasks are stored as an invasive single-linked list, starting
+/// from `FirstChild` and continuing through the `NextChild` pointers of all
+/// the linked children.
+///
+/// All children of the specific `Group` are stored "by" this record,
+/// so that they may be cancelled when this task becomes cancelled.
+///
+/// When the group exits, it may simply remove this single record from the task
+/// running it. As it has guaranteed that the tasks have already completed.
+///
+/// Group child tasks DO NOT have their own `ChildTaskStatusRecord` entries,
+/// and are only tracked by their respective `TaskGroupTaskStatusRecord`.
```

Comment:
We had a brief chat about those at some point, from which my understanding was that "one off" children would just create a `ChildTaskStatusRecord` for each and the group would be this link.

I'm totally happy to adjust either way but perhaps in a follow up PR. The complete design of child tasks isn't clear to me still... there are ideas floating around to express them just as a group, a few async lets could be in the same group perhaps etc.. But a group's way of handling cancellation then comes in the way... since it will reject adds when it is cancelled (according to present day proposal at least) which would not work for async lets which can't just magically "not run".

So I don't know here... Happy to do what you would think is best in a follow up.

Reply:
Note that the `cancelAll` needs to cancel "all children of this group" but it does _not_ cancel this task and other children of it.

So:

```swift
async let x = ... 
await Task.withGroup { 
  ... some child tasks ...
  group.cancelAll()
}

// x is not cancelled
// task is not cancelled
```

so it is useful to have "all children of this group" in a list easily accessible, so the cancelAll can just go over it. Otherwise, if we stored all children in the `task` as this linked `ChildTaskStatusRecord` it becomes difficult to "cancel only those from this group", although still possible of course... because each such child will have a `TaskGroupChildFragment` so we could iterate over all children, and cancel only those which belong to the now being cancelled group.

Either's doable. I did not quite like having to iterate over other tasks for the cancelAll so the current shape came to be.